File: src/Butil/Bit.Butil/Publics/Bluetooth.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/Web_Bluetooth_API">Web Bluetooth API</see>
/// (<c>navigator.bluetooth</c>): pick a Bluetooth Low Energy device, connect to its GATT server, and
/// read, write or subscribe to its characteristics.
/// </summary>
/// <remarks>
/// Chromium only, and only over HTTPS. <see cref="RequestDevice"/> opens the browser's device
/// chooser and so must run inside a user gesture; the grant it produces covers exactly the services
/// the request named, which is why <see cref="BluetoothRequestOptions.OptionalServices"/> is
/// usually the property that decides whether a device is usable.
/// </remarks>
[ButilService(typeof(Bluetooth))]
public class Bluetooth(IJSRuntime js) : IAsyncDisposable
{
    internal const string ValueChangedMethodName = nameof(InvokeBluetoothValueChanged);
    internal const string DisconnectedMethodName = nameof(InvokeBluetoothDisconnected);

    // Keyed by subscription id, but carrying the device each one belongs to: releasing a single
    // handle has to drop that device's handlers without touching the others'.
    private readonly ConcurrentDictionary<Guid, (string DeviceId, Action<byte[]> Handler)> _valueHandlers = new();
    private readonly ConcurrentDictionary<Guid, (string DeviceId, Action Handler)> _disconnectHandlers = new();

    // Every handle this service handed out, so a scope/circuit teardown disconnects the radio even
    // when the caller never disposed the device itself.
    private readonly ConcurrentDictionary<string, BluetoothDevice> _devices = new();

    // Per-instance callback reference (see Keyboard): notifications are isolated per circuit /
    // WASM app and released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<Bluetooth>? _dotNetRef;
    private DotNetObjectReference<Bluetooth> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>navigator.bluetooth</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.bluetooth.isSupported");

    /// <summary>
    /// True when the machine has a Bluetooth adapter the browser can use. False on a supported
    /// browser whose radio is switched off, which is the case worth telling the user about before
    /// opening a chooser that would be empty.
    /// </summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> GetAvailability() => js.Invoke<bool>("BitButil.bluetooth.getAvailability");

    /// <summary>
    /// Opens the browser's device chooser and returns the device the user picked, or null when
    /// they dismissed it. Must be called from a user gesture.
    /// </summary>
    /// <param name="options">
    /// Which devices the chooser lists, and which services the page may then use. Pass null to
    /// list every nearby device - which grants no services, so it is only useful for discovery.
    /// </param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(BluetoothDeviceInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(BluetoothRequestOptions))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(BluetoothFilter))]
    public async ValueTask<BluetoothDevice?> RequestDevice(BluetoothRequestOptions? options = null)
    {
        var info = await js.Invoke<BluetoothDeviceInfo?>("BitButil.bluetooth.requestDevice",
            options ?? new BluetoothRequestOptions { AcceptAllDevices = true });

        return info is null ? null : Track(info);
    }

    /// <summary>
    /// The devices this origin has already been granted, without showing a chooser. Empty on
    /// engines that don't implement <c>getDevices</c> even though they implement the rest.
    /// </summary>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(BluetoothDeviceInfo))]
    public async ValueTask<BluetoothDevice[]> GetDevices()
    {
        var infos = await js.Invoke<BluetoothDeviceInfo[]>("BitButil.bluetooth.getDevices");
        return [.. infos.Select(Track)];
    }

    // One handle per browser-side id. bluetooth.ts idOf() deliberately hands the same id back for the
    // same device, so a device surfaced again by GetDevices has to return the handle already in play:
    // a second wrapper over one JS registry entry would let disposing either of them release the
    // device out from under the other. Info stays the snapshot the first handle was created from - it
    // is documented as not updating on its own.
    private BluetoothDevice Track(BluetoothDeviceInfo info) => _devices.GetOrAdd(info.Id, _ => new BluetoothDevice(this, js, info));

    // Called by a handle that is disposing itself, so the service stops holding it. The JS-side
    // release() detaches every listener attached through that handle, so the .NET closures behind
    // them have to go with it - otherwise disposing one device leaks the handlers of that device
    // for as long as the service lives. The registry removal is identity-checked so a handle
    // that has already been replaced cannot untrack its successor.
    internal void Forget(BluetoothDevice device)
    {
        _devices.TryRemove(new KeyValuePair<string, BluetoothDevice>(device.Id, device));

        foreach (var pair in _valueHandlers)
        {
            if (pair.Value.DeviceId == device.Id) _valueHandlers.TryRemove(pair.Key, out _);
        }

        foreach (var pair in _disconnectHandlers)
        {
            if (pair.Value.DeviceId == device.Id) _disconnectHandlers.TryRemove(pair.Key, out _);
        }
    }

    /// <summary>
    /// Invoked from JS when a subscribed characteristic's value changes. Public +
    /// <see cref="JSInvokableAttribute"/> so it can be dispatched through the per-instance
    /// <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(ValueChangedMethodName)]
    public void InvokeBluetoothValueChanged(Guid id, byte[] value)
    {
        if (_valueHandlers.TryGetValue(id, out var entry)) entry.Handler.Invoke(value);
    }

    /// <summary>
    /// Invoked from JS on <c>gattserverdisconnected</c>. See <see cref="InvokeBluetoothValueChanged"/>.
    /// </summary>
    [JSInvokable(DisconnectedMethodName)]
    public void InvokeBluetoothDisconnected(Guid id)
    {
        if (_disconnectHandlers.TryGetValue(id, out var entry)) entry.Handler.Invoke();
    }

    [DynamicDependency(nameof(InvokeBluetoothValueChanged), typeof(Bluetooth))]
    internal async ValueTask<ButilSubscription> SubscribeValueChanged(string deviceId, string serviceUuid, string characteristicUuid, Action<byte[]> handler)
    {
        return await ButilSubscriptionHelper.Register(_valueHandlers, (deviceId, handler),
                                                      id => js.InvokeRegister("BitButil.bluetooth.startNotifications", id, DotNetRef, deviceId, serviceUuid, characteristicUuid),
                                                      id => js.InvokeVoid("BitButil.bluetooth.stopNotifications", id),
                                                      $"Notifications could not be started for characteristic '{characteristicUuid}' of service '{serviceUuid}'.");
    }

    [DynamicDependency(nameof(InvokeBluetoothDisconnected), typeof(Bluetooth))]
    internal async ValueTask<ButilSubscription> SubscribeDisconnected(string deviceId, Action handler)
    {
        return await ButilSubscriptionHelper.Register(_disconnectHandlers, (deviceId, handler),
                                                      id => js.InvokeRegister("BitButil.bluetooth.subscribeDisconnect", id, DotNetRef, deviceId),
                                                      id => js.InvokeVoid("BitButil.bluetooth.unsubscribeDisconnect", id),
                                                      "The disconnect listener could not be attached - the device handle is no longer known.");
    }

    /// <summary>
    /// Disconnects every device this service handed out, detaches its listeners and releases the
    /// interop reference.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            // Releasing a device stops its notifications and detaches its disconnect listener, and
            // every entry left here belongs to a device still in _devices - Forget() strips the rest -
            // so the loop below covers them without a round-trip per subscription.
            _valueHandlers.Clear();
            _disconnectHandlers.Clear();

            foreach (var device in _devices.Values.ToArray())
            {
                await device.DisposeAsync();
            }
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }

        GC.SuppressFinalize(this);
    }
}

File: src/Butil/Bit.Butil/Scripts/bluetooth.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    // A BluetoothDevice object *is* the permission grant the user made in the chooser, so it never
    // crosses the interop boundary - .NET holds the Guid it minted for the device and every call
    // comes back through this registry. Nothing here is reachable without a prior requestDevice()
    // or getDevices(), which is what keeps the grant scoped.
    const _devices: { [id: string]: any } = {};

    // 'characteristicvaluechanged' and 'gattserverdisconnected' listeners, keyed by the
    // subscription id .NET minted for each.
    const _notifications: { [id: string]: { device: any, characteristic: any, handler: EventListener } } = {};
    const _disconnects: { [id: string]: { device: any, handler: EventListener } } = {};

    // Handle ids are minted here rather than by .NET: getDevices() surfaces devices .NET has never
    // seen, so there is nothing for it to key them on until the info comes back. Every entry point
    // that hands a device out therefore hands out its id with it.
    let _sequence = 0;
    function nextId() { return `bt${++_sequence}`; }

    butil.bluetooth = {
        isSupported() { return !!(navigator as any).bluetooth; },
        getAvailability,
        requestDevice,
        getDevices,
        forget,
        release,
        connect,
        disconnect,
        isConnected,
        getPrimaryServices,
        getCharacteristics,
        readValue,
        writeValue,
        startNotifications,
        stopNotifications,
        subscribeDisconnect,
        unsubscribeDisconnect
    };

    function bluetooth() { return (navigator as any).bluetooth; }

    function describe(id: string, device: any) {
        return {
            id,
            deviceId: device.id ?? '',
            name: device.name ?? null,
            connected: !!device.gatt?.connected
        };
    }

    // A 16-bit alias ("0x180d"), a name ("heart_rate") and a full UUID are all legal service /
    // characteristic identifiers, and all three arrive as strings from .NET. Hex digits are passed
    // through as a number so the browser resolves the alias itself.
    function uuid(value: string): any {
        if (typeof value !== 'string') return value;
        const trimmed = value.trim();
        if (/^0x[0-9a-f]+$/i.test(trimmed)) return parseInt(trimmed, 16);
        return trimmed;
    }

    function bytes(view: DataView) {
        return new Uint8Array(view.buffer, view.byteOffset, view.byteLength);
    }

    async function getAvailability() {
        const bt = bluetooth();
        if (!bt || typeof bt.getAvailability !== 'function') return false;
        try { return await bt.getAvailability(); } catch { return false; }
    }

    async function requestDevice(options: any) {
        const bt = bluetooth();
        if (!bt) return null;

        // acceptAllDevices and filters are mutually exclusive - passing both is a TypeError - and
        // acceptAllDevices is documented to win, so the filters aren't even looked at then.
        const acceptAll = options?.acceptAllDevices === true;
        const supplied = acceptAll ? [] : (options?.filters ?? []);

        const filters = supplied.map((filter: any) => {
            const out: any = {};
            if (filter.services?.length) out.services = filter.services.map(uuid);
            if (filter.name) out.name = filter.name;
            if (filter.namePrefix) out.namePrefix = filter.namePrefix;
            return out;
        }).filter((filter: any) => Object.keys(filter).length > 0);

        // Filters that were asked for but carry no criterion are a mistake in the caller's options.
        // Silently widening that to "every device nearby" is the opposite of what they asked for.
        if (supplied.length && filters.length === 0)
            throw new TypeError('Every supplied Bluetooth filter is empty - give each one a service, name or namePrefix, or set AcceptAllDevices.');

        const request: any = {};
        if (filters.length) request.filters = filters;
        else request.acceptAllDevices = true;
        if (options?.optionalServices?.length) request.optionalServices = options.optionalServices.map(uuid);

        let device: any;
        try {
            device = await bt.requestDevice(request);
        } catch (e: any) {
            // NotFoundError is what dismissing the chooser looks like - "no device", not a failure.
            // Everything else (a SecurityError for a service the request never named, a TypeError
            // for a malformed filter) is a mistake in the calling code and has to reach it.
            if (e?.name === 'NotFoundError') return null;
            throw e;
        }

        const id = nextId();
        _devices[id] = device;
        return describe(id, device);
    }

    // Devices the user has already permitted, each under a fresh handle id.
    async function getDevices() {
        const bt = bluetooth();
        if (!bt || typeof bt.getDevices !== 'function') return [];
        try {
            const devices = await bt.getDevices();
            return devices.map((device: any) => {
                const id = nextId();
                _devices[id] = device;
                return describe(id, device);
            });
        } catch { return []; }
    }

    async function forget(id: string) {
        const device = _devices[id];
        if (!device || typeof device.forget !== 'function') return false;
        try { await device.forget(); release(id); return true; } catch { return false; }
    }

    // Drops the registry entry and every listener attached through it. Called on handle disposal;
    // the device itself stays permitted - forget() is what revokes the grant.
    function release(id: string) {
        const device = _devices[id];
        delete _devices[id];
        if (!device) return;

        for (const key of Object.keys(_disconnects)) {
            if (_disconnects[key].device === device) unsubscribeDisconnect(key);
        }
        // Notifications are attached through the same handle, so they go with it. Fire-and-forget:
        // stopNotifications drops the entry and the listener synchronously, and only the device
        // round-trip is left to settle on its own.
        for (const key of Object.keys(_notifications)) {
            if (_notifications[key].device === device) void stopNotifications(key).catch(() => { /* device already gone */ });
        }
        try { device.gatt?.disconnect(); } catch { /* already gone */ }
    }

    async function connect(id: string) {
        const device = _devices[id];
        if (!device?.gatt) return false;
        try { await device.gatt.connect(); return true; } catch { return false; }
    }

    function disconnect(id: string) {
        try { _devices[id]?.gatt?.disconnect(); } catch { /* already gone */ }
    }

    function isConnected(id: string) { return !!_devices[id]?.gatt?.connected; }

    async function server(id: string) {
        const device = _devices[id];
        if (!device?.gatt) return null;
        // A GATT server drops the connection between operations more often than not, so every
        // operation re-connects rather than assuming an earlier connect() still holds.
        if (!device.gatt.connected) await device.gatt.connect();
        return device.gatt;
    }

    async function getPrimaryServices(id: string) {
        const gatt = await server(id);
        if (!gatt) return [];
        const services = await gatt.getPrimaryServices();
        return services.map((service: any) => ({ uuid: service.uuid, isPrimary: !!service.isPrimary }));
    }

    async function getCharacteristics(id: string, serviceUuid: string) {
        const gatt = await server(id);
        if (!gatt) return [];
        const service = await gatt.getPrimaryService(uuid(serviceUuid));
        const characteristics = await service.getCharacteristics();
        return characteristics.map((characteristic: any) => ({
            uuid: characteristic.uuid,
            serviceUuid: service.uuid,
            broadcast: !!characteristic.properties?.broadcast,
            read: !!characteristic.properties?.read,
            writeWithoutResponse: !!characteristic.properties?.writeWithoutResponse,
            write: !!characteristic.properties?.write,
            notify: !!characteristic.properties?.notify,
            indicate: !!characteristic.properties?.indicate
        }));
    }

    async function characteristic(id: string, serviceUuid: string, characteristicUuid: string) {
        const gatt = await server(id);
        if (!gatt) return null;
        const service = await gatt.getPrimaryService(uuid(serviceUuid));
        return await service.getCharacteristic(uuid(characteristicUuid));
    }

    async function readValue(id: string, serviceUuid: string, characteristicUuid: string) {
        const target = await characteristic(id, serviceUuid, characteristicUuid);
        if (!target) return null;
        const value = await target.readValue();
        return bytes(value);
    }

    async function writeValue(id: string, serviceUuid: string, characteristicUuid: string, data: Uint8Array, withResponse: boolean) {
        const target = await characteristic(id, serviceUuid, characteristicUuid);
        if (!target) return false;
        const buffer = butil.utils.arrayToBuffer(data);
        // writeValueWithResponse/WithoutResponse are the current spec; writeValue is the legacy
        // name still shipping in older Chromium, and the only one some builds expose.
        if (withResponse && typeof target.writeValueWithResponse === 'function') await target.writeValueWithResponse(buffer);
        else if (!withResponse && typeof target.writeValueWithoutResponse === 'function') await target.writeValueWithoutResponse(buffer);
        else await target.writeValue(buffer);
        return true;
    }

    async function startNotifications(subscriptionId: string, dotNetRef: any, id: string, serviceUuid: string, characteristicUuid: string) {
        const target = await characteristic(id, serviceUuid, characteristicUuid);
        if (!target) return false;

        const handler: EventListener = event => {
            const value = (event.target as any).value as DataView;
            butil.utils.dispatch(dotNetRef, 'InvokeBluetoothValueChanged', subscriptionId, bytes(value));
        };
        // Attached before starting so the first notification can't slip through, and taken back off
        // if the device refuses - a rejected start would otherwise leave a listener nothing tracks
        // and nothing can remove.
        target.addEventListener('characteristicvaluechanged', handler);
        try {
            await target.startNotifications();
        } catch (e) {
            target.removeEventListener('characteristicvaluechanged', handler);
            throw e;
        }
        _notifications[subscriptionId] = { device: _devices[id], characteristic: target, handler };
        return true;
    }

    async function stopNotifications(subscriptionId: string) {
        const entry = _notifications[subscriptionId];
        if (!entry) return;
        delete _notifications[subscriptionId];
        entry.characteristic.removeEventListener('characteristicvaluechanged', entry.handler);
        try { await entry.characteristic.stopNotifications(); } catch { /* device already gone */ }
    }

    function subscribeDisconnect(subscriptionId: string, dotNetRef: any, id: string) {
        const device = _devices[id];
        if (!device) return false;
        const handler: EventListener = () => butil.utils.dispatch(dotNetRef, 'InvokeBluetoothDisconnected', subscriptionId);
        device.addEventListener('gattserverdisconnected', handler);
        _disconnects[subscriptionId] = { device, handler };
        return true;
    }

    function unsubscribeDisconnect(subscriptionId: string) {
        const entry = _disconnects[subscriptionId];
        if (!entry) return;
        delete _disconnects[subscriptionId];
        entry.device.removeEventListener('gattserverdisconnected', entry.handler);
    }
}(BitButil));
